### ISSUE TITLE
Question about updating the domain classifier parameters

### ISSUE THREAD

**Issue by brijmohan** [fungtion/DANN#7]
Hi @fungtion ,

I am trying to implement a similar architecture for speech processing and I am using your code for reference.

According to my understanding, DANN must minimize the loss of class labels and maximize the loss of domain with respect to the features. It must also optimize just the parameters of domain classifier to make it a good "adversary" by minimizing the domain loss without affecting the feature layer params.

How do you achieve the optimization of just the domain classifier parameters without propagating the gradients back to the feature extraction layer? Could you please point me towards the code that does it?

Thanks a lot for providing your helpful code.

Brij

**Comment by fungtion** [comment]
In my opinion, DANN minimize the loss of class labels,  and prevent the minimizing of the loss of domain, which is not equal to maximizing it.  All those purposes are achieved by negtive adversarial layer, where the negtive gradients backpropagated from the loss of doamin. 
The key point of domain adaption methods is to force the model to extract the common features of the data from different domains. For the datasets minist and mnist_m, preventing the minimizing of the loss of domain can force the feature layer to extract the comman features of them, i.e. digits, and to filter these irrevant features, i.e. colorful background of the digits in mnist_m. So, the positive gradient from the the loss of  label and the negtive gradient from the loss of domain can work together to achieve doamin adaption.

**Comment by qleroy** [comment]
Hi @brijmohan ,
The piece of code that makes the desired parameter updates possible is the `ReverseLayerF` Function located in [models/functions.py](https://github.com/fungtion/DANN/blob/master/models/functions.py#L4). It implements the so-called "Gradient Reversal Layer" depicted in the teaser figure of the README and explained in the linked paper. 
`ReverseLayerF` is applied in [models/model.py](https://github.com/fungtion/DANN/blob/master/models/model.py#L42) forward(self, input_data, alpha) function. 
See also the "extending pytorch" [note](https://pytorch.org/docs/stable/notes/extending.html) to understand the implementation.

**Comment by brijmohan** [comment]
Hi @fungtion @qleroy ,

Thanks for the clear explanation. If I understood correctly, 

> The "Gradient reversal layer" indeed: optimizes the **domain branch** to minimize its loss + also "reverses" these gradients at the **feature extraction layer**, so that the learned representation is invariant of domain. Minimizing the class label loss optimizes the desired classifier branch.


These two losses are sufficient for domain adversarial training. I do not need to explicitly train the domain branch to be a good "adversary". It is implicit in the above training procedure.

Could you please comment if this understanding is correct?

Thanks a lot!

**Comment by fungtion** [comment]
@brijmohan Yes, that's right.

**Comment by brijmohan** [comment]
Okay! Nice to hear that :)
